feat(gce): document how to deploy Pro Kubernetes cluster on GCE (#28)
canonical/ubuntu-cloud-docs: google/gke/gke-how-to/deploy-pro-gke-cluster.rst

Deploy a Pro GKE Cluster
========================

Limitations
-----------

Google does not yet have Ubuntu Pro image offerings for the GKE nodes i.e. you cannot choose
Ubuntu Pro images for the GKE nodes. GKE does not support custom images for the nodes either.
Additionally, GKE does not allow for post-deployment customisation of node VMs.

	"Modifications on the boot disk of a node VM do not persist across node re-creations.
	Nodes are re-created during manual upgrade, auto-upgrade, auto-repair, and auto-scaling.
	In addition, nodes are re-created when you enable a feature that requires node re-creation,
	such as GKE Sandbox, intranode visibility, and shielded nodes."

	-- `GKE docs <https://cloud.google.com/kubernetes-engine/docs/concepts/node-images#modifications>`_

Given that, currently, there's no mechanism to either enable Ubuntu Pro or pre-bake the UA token

> in the specific cluster, so the managed Kubernetes is not applicable for the GCE.

Reviewer: ```suggestion
in the specific cluster, so the managed Pro Kubernetes cluster is not applicable in GCE.
```

Deploy Kubernetes manually on Ubuntu Pro VMs
--------------------------------------------

> One solution to have a Ubuntu Pro cluster is to deploy and manage Kubernetes manually on

Reviewer: ```suggestion
One solution to have an Ubuntu Pro cluster is to deploy and manage Kubernetes manually on
```

> Ubuntu Pro VMs available in GCP.

Reviewer: ```suggestion
Ubuntu Pro VMs available in GCE.
```

GCP was correct, but I think GCE makes as much sense here and it avoids using too many G** acronyms

Create Ubuntu Pro VMs for the control plane and worker nodes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Create a few Ubuntu Pro VMs for the worker nodes and control plane from the
`GCE VM Instances <https://console.cloud.google.com/compute/instances>`_ page.
There is a `Create Instance <https://console.cloud.google.com/compute/instancesAdd>`_ button
which takes you to the configuration details before creating the VM.

Make sure to change the boot disk configuration so that the operating system

> is "Ubuntu Pro". There are a few versions to choose from:

Reviewer: ```suggestion
is "Ubuntu Pro". There are a few versions to choose from, such as:
```

(there could be more in the future)

- Ubuntu 16.04 LTS Pro Server
- Ubuntu 18.04 LTS Pro Server
- Ubuntu 20.04 LTS Pro Server
- Ubuntu 22.04 LTS Pro Server
- Ubuntu 18.04 LTS Pro FIPS Server
- Ubuntu 20.04 LTS Pro FIPS Server

Choose the version suitable for your needs, and create the VMs.

You can also utilise the `gcloud CLI tool <https://cloud.google.com/sdk/gcloud>`_ to deploy the VMs
with the following command.

::

	$ gcloud compute instances create <instance-name> <options..>

As an example, you can create two Ubuntu Pro VMs for the worker nodes named ``k8s-worker-1``
and ``k8s-worker-2`` and another Pro VM for the control plane named ``k8s-main``, each of them

> on "Ubuntu Pro" operating system with "Ubuntu 20.04 LTS Pro FIPS Server" version.

Reviewer: ```suggestion
```

You can access the VMs via SSH with the following command.

::

	$ gcloud compute ssh --zone <instance-zone> <instance-name> --project <project-name>

Install Kubernetes and create a cluster
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Install Kubernetes on the nodes to create and manage the cluster. You can use
`MicroK8s <https://microk8s.io/>`_ to meet your Kubernetes needs. Install the snap.

::

> 

Reviewer: ```suggestion

	# repeat for each node
```

	$ sudo snap install microk8s --classic

To create a cluster out of two or more already-running MicroK8s instances, use the
``microk8s add-node`` command. The MicroK8s instance on which this command is run

> will be the master of the cluster and will host the Kubernetes control plane.

Reviewer: ```suggestion
will be the cluster's manager and will host the Kubernetes control plane.
```

> Following the example VMs, run the command in ``k8s-main``.

Reviewer: ```suggestion
Following the exemplary VMs from above, run the following command in ``k8s-main``.
```

::

	root@k8s-main:~# microk8s add-node
	From the node you wish to join to this cluster, run the following:
	microk8s join 10.128.0.7:25000/3a9547446eae54765ecc0bc522321968/079215076aa1

	Use the '--worker' flag to join a node as a worker not running the control plane, eg:
	microk8s join 10.128.0.7:25000/3a9547446eae54765ecc0bc522321968/079215076aa1 --worker

	If the node you are adding is not reachable through the default interface you can use one of the following:
	microk8s join 10.128.0.7:25000/3a9547446eae54765ecc0bc522321968/079215076aa1

Note that, the IP address, port, and token generated here will be project specific.
You will not have the same values as above. Find more info in the
`microk8s clustering doc <https://microk8s.io/docs/clustering>`_.

> In the worker nodes, execute the respective command to join the cluster as a worker.

Reviewer: ```suggestion
Then, SSH into each of the worker nodes and execute the respective command to join the cluster as a worker.
```

::

> 

Reviewer: ```suggestion

	# repeat for each worker node
```

		/var/snap/microk8s/current/args/traefik/provider.yaml

In the control plane VM, you should be able to see that the nodes have joined the cluster using
the ``kubetl get nodes`` command.

::

	root@k8s-main:~# microk8s kubectl get nodes --output=wide
	NAME           STATUS   ROLES    AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION        CONTAINER-RUNTIME
	k8s-worker-2   Ready    <none>   2m13s   v1.26.1   10.128.0.10   <none>        Ubuntu 20.04.5 LTS   5.4.0-1021-gcp-fips   containerd://1.6.8
	k8s-main       Ready    <none>   78m     v1.26.1   10.128.0.7    <none>        Ubuntu 20.04.5 LTS   5.4.0-1021-gcp-fips   containerd://1.6.8
	k8s-worker-1   Ready    <none>   16m     v1.26.1   10.128.0.8    <none>        Ubuntu 20.04.5 LTS   5.4.0-1021-gcp-fips   containerd://1.6.8

You can also check out the cluster-info using the ``kubectl cluster-info`` command.

::

	root@k8s-main:~# microk8s kubectl cluster-info
	Kubernetes control plane is running at https://127.0.0.1:16443

	To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

Conclusion
~~~~~~~~~~

> You now have a Ubuntu Pro Kubernetes cluster running on Ubuntu Pro nodes in GCP.

Reviewer: Can you please add a one-liner to show **how** we know we are running in a Pro cluster? I guess you can use either the `ua` or `pro` CLI clients to assert that. So something like:

```suggestion
You now have an Ubuntu Pro Kubernetes cluster running in GCE. Your Pro subscription can be verified on each of the provisioned nodes by running:

#### TODO: 

And, from the control plane node (``k8s-main`` in the example above), you'll also find the _KUBECONFIG_ files at ``TODO``, which you can then use to control your Pro Kubernetes cluster from any working environment with a Kubernetes client.
```

Author: Done.
